review/get_review_ids.py

```diff
@@ -0,0 +1,235 @@
+import pytz
+from courseware.models import StudentModule
+from enrollment.api import add_enrollment, get_enrollment, update_enrollment
+from lms.djangoapps.course_blocks.api import get_course_blocks
+from xmodule.modulestore.django import modulestore
+
+from .configuration import ENROLLMENT_COURSE_MAPPING, REVIEW_COURSE_MAPPING, TEMPLATE_URL
+
+log = logging.getLogger(__name__)
+
+# TODO: Switch to using CourseLocators and/or CourseKeys everywhere
+
+
+def get_problems(num_desired, current_course):
+    '''
+    Looks through all the problems a learner has previously loaded and randomly
+    selects num_desired of them. Also checks if the learner had originally
+    answered it correctly or incorrectly and after how many attempts.
+
+    Parameters:
+        num_desired (int): the number of desired problems to show the learner
+        current_course (CourseLocator): The course the learner is currently in
+
+    Returns a list of num_desired tuples in the form (URL to display, correct, attempts)
+    '''
+    user = crum.get_current_user()
+
+    enroll_user_in_review_course(user, current_course)
+    store = modulestore()
+
+    problem_data = []
+
+    for block_key, state in get_records(user, current_course):
+        if is_valid_problem(store, block_key, state):
+            correct, attempts = get_correctness_and_attempts(state)
+            problem_id = block_key.block_id
+            problem_data.append((problem_id, correct, attempts))
+            delete_state_of_review_problem(user, current_course, problem_id)
+
+    if len(problem_data) < num_desired:
+        return []
+
+    problems_to_show = random.sample(problem_data, num_desired)
+    review_course_id = REVIEW_COURSE_MAPPING[str(current_course)]
+    problem_information = []
+    for problem, correct, attempts in problems_to_show:
+        problem_information.append((TEMPLATE_URL.format(course_id=review_course_id,
+                                    type='problem', xblock_id=problem), correct, attempts))
+    return problem_information
+
+
+def get_vertical(current_course):
+    '''
+    Looks through all the problems a learner has previously loaded and
+    finds their parent vertical. Then randomly selects a single vertical
+    to show the learner.
+
+    Parameters:
+        current_course (CourseLocator): The course the learner is currently in
+
+    Returns a url (str) with the vertical id to render for review.
+    '''
+    user = crum.get_current_user()
+
+    enroll_user_in_review_course(user, current_course)
+
+    store = modulestore()
+    course_usage_key = store.make_course_usage_key(current_course)
+    course_blocks = get_course_blocks(user, course_usage_key)
+
+    vertical_data = set()
+
+    for block_key, state in get_records(user, current_course):
+        if is_valid_problem(store, block_key, state):
+            parent = course_blocks.get_parents(block_key)[0]
+            vertical_id = parent.block_id
+            vertical_data.add(vertical_id)
```

Comment:
Question: would a "valid vertical" be a vertical that has a maximum number of blocks (maybe 5).  Otherwise, do want learners to review very long verticals with many blocks?

Reply:
At this point in time, all verticals are valid since they are just coming from the course. The reason of displaying the verticals is because all of the problems inside of it are dependent on one another, so I don't believe course teams will go overboard with the size of these. Ideally, verticals that are rather large (say > 5) contain distinct problems in which case we would display the review material by a number of desired problems rather than a single vertical.